homeassistant/components/camera/synology.py

```diff
@@ -0,0 +1,209 @@
+    CAMERA_PATH = query_resp['data'][CAMERA_API]['path']
+    STREAMING_PATH = query_resp['data'][STREAMING_API]['path']
+
+    # Authticate to NAS to get a session id
+    syno_auth_url = SYNO_API_URL.format(config.get(CONF_URL),
+                                        WEBAPI_PATH,
+                                        AUTH_PATH)
+
+    auth_payload = {'api': AUTH_API,
+                    'method': 'Login',
+                    'version': '2',
+                    'account': config.get(CONF_USERNAME),
+                    'passwd': config.get(CONF_PASSWORD),
+                    'session': 'SurveillanceStation',
+                    'format': 'sid'}
+    auth_req = requests.get(syno_auth_url,
+                            params=auth_payload,
+                            verify=config.get(CONF_VALID_CERT),
+                            timeout=TIMEOUT)
+    auth_resp = loads(auth_req.text)
+    session_id = auth_resp['data']['sid']
+
+    # Use SessionID to get cameras in system
+    syno_camera_url = SYNO_API_URL.format(config.get(CONF_URL),
+                                          WEBAPI_PATH,
+                                          camera_api)
+    camera_payload = {'api': CAMERA_API,
+                      'method': 'List',
+                      'version': '1'}
+    camera_req = requests.get(syno_camera_url,
+                              params=camera_payload,
+                              verify=config.get(CONF_VALID_CERT),
+                              timeout=TIMEOUT,
+                              cookies={'id': session_id})
+    camera_resp = loads(camera_req.text)
+    cameras = camera_resp['data']['cameras']
+    for camera in cameras:
+        if not config.get(CONF_WHITELIST):
+            camera_id = camera['id']
+            snapshot_path = camera['snapshot_path']
+            add_devices([SynologyCamera(config,
+                                        camera_id,
+                                        camera['name'],
+                                        snapshot_path)])
+
+
+# pylint: disable=too-many-instance-attributes
+class SynologyCamera(Camera):
+    """An implementation of a Synology NAS based IP camera."""
+
+    def __init__(self, config, camera_id, camera_name, snapshot_path):
+        """Initialize a Synology Surveillance Station camera."""
+        super().__init__()
+        self._name = camera_name
+        self._username = config.get(CONF_USERNAME)
+        self._password = config.get(CONF_PASSWORD)
+        self._synology_url = config.get(CONF_URL)
+        self._api_url = config.get(CONF_URL) + 'webapi/'
+        self._login_url = config.get(CONF_URL) + '/webapi/' + 'auth.cgi'
+        self._camera_name = config.get(CONF_CAMERA_NAME)
+        self._stream_id = config.get(CONF_STREAM_ID)
+        self._valid_cert = config.get(CONF_VALID_CERT)
+        self._camera_id = camera_id
+        self._snapshot_path = snapshot_path
+
+        self.get_sid()
+
+    def get_sid(self):

```

Comment:
This method is also used in the setup. I would suggest to make it a helper function

```python
def get_sid(username, password, conf_url, verify_ssl)
```